RTD Rapid Tiling Device

SyncIt21 opened the pull request:
## About The Pull Request
Interior decor at its finest & fastest available in engineering Proto lathe after research 
- Left click on a tile to infer its type. If the tile is supported by the RTD then a balloon alert will be displayed showing "Tile changed to XXX". If not supported, it will tell you that
- Right click on tile to convert it back to plating. It can only deconstruct floor types which are supported
- Left click on plating to place your floor tile
- Only iron & glass floor types are supported because you can only feed iron & glass to the RTD to keep it consistent

Also, this is my first time ever making sprites so expect less.

https://user-images.githubusercontent.com/110812394/209545438-6a51e7bf-163a-4a18-9102-7b77107eb1b7.mp4

Cleaned up some code in the RCD file as a bonus some notables ones are
- moved `update_overlays()` from rcd to `obj/item/construction `so any class extending from this can use it
- removed `do_after()` when changing turf types cause its unessassary & doesn't play well with the RTD

## Why It's Good For The Game
- Replace damaged or dirty tiles without the need of a janitor, crowbar & removes manual labour by hand
- Give your room some style quickly maybe replace the whole station floor with glass tiles I don't know you decide
- Maybe lay floor tiles quickly in maintenance to stop rats from chewing on those cables

## Changelog
:cl:
add: RTD Rapid Tiling Device
imageadd: Sprite for RTD
code: moved update_overlays() from rcd to obj/item/construction so anyone can use it 
del: cooldown of 3 seconds when changing turf type
qol: Faster Turf decoration
/:cl:

SyncIt21 committed:
RTD Rapid Tiling Device

SyncIt21 committed:
Spelling

SyncIt21 committed:
Merge branch 'master' into RTD

SyncIt21 committed:
Removed duplicate

SyncIt21 committed:
Merge branch 'RTD' of https://github.com/SyncIt21/The-TG-Station-Fork into RTD

SyncIt21 committed:
Duplicate

NamelessFairy commented:
Just tested on local, found a bug regarding reinforced plating, if you put a floor tile ontop of reinforced plating and then remove it with the RTD it removes the floor tile and the reinforced plating leaving only the regular plating below

SyncIt21 commented:
> Just tested on local, found a bug regarding reinforced plating, if you put a floor tile ontop of reinforced plating and then remove it with the RTD it removes the floor tile and the reinforced plating leaving only the regular plating below

I see . Ill take a look

NamelessFairy commented:
> I see . I will take a look

Your probably looking for the ScrapAway() proc by the way, its a proc on all turfs that removes the top layered turf and replaces it with the layer below.

SyncIt21 commented:
> > I see . I will take a look
> 
> Your probably looking for the ScrapAway() proc by the way, its a proc on all turfs that removes the top layered turf and replaces it with the layer below.

The best solution is to not allow the RTD to put floors on reinforced plating because its not possible to remember what the previous plating was before replacing it. I'll do that

SyncIt21 commented:
> Scr

Sorry forget my previous comment. I'll try your solution first

SyncIt21 commented:
It seems scrap away creates a problem for glass turfs. It removes the plating entirely exposing space. Looks like i need to check for this specific case

SyncIt21 commented:
Ok so the problem was glass turfs have open space as the base turf thus scraping it away exposes space which is very bad. 
We have to check specifically for this case .

Fikou commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,359 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+				new /datum/tile_info("Smooth Large", /obj/item/stack/tile/iron/dark/smooth_large, 7),
+				new /datum/tile_info("Small", /obj/item/stack/tile/iron/dark/small, 4),
+				new /datum/tile_info("Diagonal", /obj/item/stack/tile/iron/dark/diagonal, 4),
+				new /datum/tile_info("Herrigbone", /obj/item/stack/tile/iron/dark/herringbone, 4),
```
herringbone

SyncIt21 replied:
Sorry

SyncIt21 committed:
Scrap Away & Change Turf

HerringBone

SyncIt21 commented:
> Just tested on local, found a bug regarding reinforced plating, if you put a floor tile ontop of reinforced plating and then remove it with the RTD it removes the floor tile and the reinforced plating leaving only the regular plating below

Problem solved

private-tristan commented:
left click for changing the tile type seems a bit strange because you'd need to go into the menu if you accidentally misclicked, maybe change it to control click?

SyncIt21 commented:
> left click for changing the tile type seems a bit strange because you'd need to go into the menu if you accidentally misclicked, maybe change it to control click?

Sure i'll do that

Mothblocks commented on code/modules/asset_cache/assets/rtd.dm:
```diff
@@ -0,0 +1,117 @@
+
+	for(var/obj/item/stack/tile/tile as anything in tiles)
+		var/icon_state = initial(tile.icon_state)
+		Insert(sprite_name = icon_state , I = 'icons/obj/tiles.dmi', icon_state = icon_state)
```
```suggestion
		Insert(sprite_name = icon_state, I = 'icons/obj/tiles.dmi', icon_state = icon_state)
```

Mothblocks commented on code/modules/asset_cache/assets/rtd.dm:
```diff
@@ -0,0 +1,117 @@
+					var/icon/east_rotated = icon(icon = 'icons/obj/tiles.dmi', icon_state = icon_state)
+					east_rotated.Turn(-90)
+					img.Blend(east_rotated,ICON_MULTIPLY)
+					img.SetIntensity(2,2,2)
```
What is SetIntensity?

SyncIt21 replied:
After blending together 2 icons the final image becomes way too dark. So I'm just scaling the r,g,b intensities of the final image so it becomes  a little brighter for the eye

Mothblocks replied:
Why are we blending multiple icons on top of each other anyway? That seems super whack, I would expect that this would just need to be rotating the tile, not blending multiple things on top of each other.

SyncIt21 replied:
It is wack ><. I can easily oriented image using the image() proc but I can't convert it into an icon for the sprite sheet. 
I don't know what to do

SyncIt21 replied:
Same problem for the below comment

Mothblocks commented on code/modules/asset_cache/assets/rtd.dm:
```diff
@@ -0,0 +1,117 @@
+					west_rotated.Turn(90)
+					img.Blend(west_rotated,ICON_MULTIPLY)
+					img.SetIntensity(2,2,2)
+			Insert(sprite_name = "[icon_state]-[dir2text(direction)]", I = img)
```
Agh, this is a lot of directions, and icon creation is not at all cheap. Ideally we would be doing this styling in the CSS instead, but I don't know if IE supports that well.

Mothblocks replied:
Actually since we already can use transform I'm fairly sure we can. Please try and see if you can get that to work.

SyncIt21 replied:
Actually I need ur advice on this.
I can create an image which will automatically orient the icon in any specified direction using image() function but I can't convert it into an icon for the sprite sheet.

How do I convert an image into an icon?

Mothblocks replied:
I'm a bit confused.

The way I would expect this to work is that this sends the tile image for every single tile, and then the client uses `rotate` in `transform` to perform the rotation based on direction.

You shouldn't need `image` at all and you can just use `icon` for sending the initial icon.

SyncIt21 replied:
Rotate accepts angle in degrees as the parameter not directions likes NORTH, SOUTH etc.

Mothblocks replied:
plz stop marking these as resolved just because you think they can't be helped, this is a seriously pressing issue that needs my actual say on

SyncIt21 commented on code/__DEFINES/rcd.dm:
```diff
@@ -0,0 +1,29 @@
+///A map of which tiles[singluar_name] can be rotated in what directions
+///used to create rotated asset icons & to display ui in rtd
+///this is required because initial() returns null when retriving list variables such as tile_rotate_dirs so had to manually type them
+GLOBAL_LIST_INIT(tile_rotations, list(
```
It's used in 2 places actually 
code/game/objects/items/RCD.dm
To display tile directions to the UI
code/modules/asset_cache/assets/rtd.dm
To create rotated icons for that specific tile type 
If not DEFINES I don't know how else to share this list

Mothblocks replied:
code/_globalvars

SyncIt21 replied:
I'll do that

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
 	desc = "Scans the surrounding area for destruction. Scanned structures will rebuild significantly faster."
 
+/obj/item/construction/rtd
+	name = "rapid-tiling-device (RTD)"
```
Yup for the RCD
https://github.com/tgstation/tgstation/blob/dc650b6fba8256d9593fd330f6023c12b5688682/code/game/objects/items/RCD.dm#L288-L289

But not the RLD
https://github.com/tgstation/tgstation/blob/dc650b6fba8256d9593fd330f6023c12b5688682/code/game/objects/items/RCD.dm#L894-L895

Which format do you want to stick with?

Mothblocks replied:
weird. what you have is best then

SyncIt21 committed:
Full stop

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+		"Conventional" = list(
+			//The most common types
+			"Standard" = list(
+				new /datum/tile_info("Base", /obj/item/stack/tile/iron, 4),
```
Its the cost.  I'll put an named varg and document that

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+	icon_state = initial(type.icon_state)
+	cost = tile_cost
+
+/datum/tile_info/proc/fill_ui_data(list/data, dir)
```
I'm not following.  What do you mean exactly?

Mothblocks replied:
`dir` -> `direction` wherever it's used, we don't like to abbreviate names

SyncIt21 replied:
Oh right I'll do that

SyncIt21 committed:
Simpler ternary

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
 	name = "Destruction Scan"
 	desc = "Scans the surrounding area for destruction. Scanned structures will rebuild significantly faster."
 
+/obj/item/construction/rtd
```
Moving out

SyncIt21 committed:
Simpler ternary with isnull()

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+	. = ..()
+	update_appearance()
+
+/obj/item/construction/rtd/Initialize(mapload)
```
Fuck

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+	return tile_directions[1]
+
+/datum/tile_info/proc/get_icon_state(selected_dir)
+	return icon_state + (selected_dir != null ? "-[dir2text(selected_dir)]" : "")
```
Is github working now? I can't commit some of your suggestions for some reason.  Anyway I'll do this

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+		var/list/designs = list() //initialize all designs under this category
+		for(var/i in 1 to target_category.len)
+			var/datum/tile_info/tile_design = target_category[i]
+			designs += list(list("index" = i, "name" = tile_design.name, "icon" = tile_design.get_icon_state()))
```
Right typescript map() has i as parameter.  Need to change this for the RCD as well.  I'll do that

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+			var/datum/tile_info/tile_design = target_category[i]
+			designs += list(list("index" = i, "name" = tile_design.name, "icon" = tile_design.get_icon_state()))
+
+		data["categories"] += list(list("cat_name" = sub_category, "recipes" = designs))
```
Yup

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+
+	return TRUE
+
+/obj/item/construction/rtd/proc/is_valid_plating(atom/A)
```
OOFFF

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+	return TRUE
+
+/obj/item/construction/rtd/proc/is_valid_plating(atom/A)
+	return A.type == /turf/open/floor/plating ||  A.type == /turf/open/floor/plating/reinforced
```
We don't want it to capture subtypes actually. RTD is allowed to only lay floor tiles on these 2 specific platings cause these 2 are the base. That's what this check is for

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+/obj/item/construction/rtd/proc/is_valid_plating(atom/A)
+	return A.type == /turf/open/floor/plating ||  A.type == /turf/open/floor/plating/reinforced
+
+/obj/item/construction/rtd/afterattack(atom/A, mob/user)
```
Names

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+						selected_dir = floor.dir
+						if(!tile_design.is_valid_dir(selected_dir)) //selected_dir will mostly be SOUTH but if the tile design doesnt support it then make it null
+							selected_dir = null
+						balloon_alert(user, "Tile changed to [selected_design.name]")
```
Yup

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+						return TRUE
+
+		//could not infer floor type
+		balloon_alert(user, "Design not supported")
```
Comments and grammer

SyncIt21 committed:
0.5 seconds

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
0.2 seconds

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
No return

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+					cost = tile_design.cost
+					can_deconstruct = TRUE
+					break
+	if(!can_deconstruct || !checkResource(cost * 0.7, user)) //no ballon alert for checkResource as it already spans an alert to chat
```
Nope it's inherited from obj/item/construction

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+			if(can_deconstruct)
+				break
+			for(var/datum/tile_info/tile_design in floor_designs[main_root][sub_category])
+				if(initial(tile_design.tile_type.turf_type) == floor.type)
```
My old nemesis

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
+
+	//All special effect stuff
+	user.Beam(A, icon_state="light_beam", time = 2)
```
Yup

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -1475,6 +1474,365 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+					break
+	if(!can_deconstruct || !checkResource(cost * 0.7, user)) //no ballon alert for checkResource as it already spans an alert to chat
+		if(!can_deconstruct)
+			balloon_alert(user, "Cannot deconstruct this type")
```
English

SyncIt21 committed:
baseturf_at_depth

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 commented on code/modules/asset_cache/assets/rtd.dm:
```diff
@@ -0,0 +1,117 @@
+/datum/asset/spritesheet/rtd
+	name = "rtd-tgui"
```
I should change this for Plumbing rcd as well

SyncIt21 commented on code/modules/asset_cache/assets/rtd.dm:
```diff
@@ -0,0 +1,117 @@
+	name = "rtd-tgui"
+
+/datum/asset/spritesheet/rtd/create_spritesheets()
+	var/list/tiles = list(
```
This list is just for convenience.  I copied all floor types from the static list in RTD and slapped it in here.
It's not possible to share this list cause inside the RTD all these types are arranged into categories and subcategories while in here its just one big linear list.
While it is duplicate on a way its not too much work for anyone to just paste their values at the end of this list

Mothblocks replied:
We don't need to share the list but we should absolutely be creating this dynamically

> Can we instead create an RTD and read its values and extract it from there instead? 

As in, we should be creating a list that loops over those values and flattens it out. Potentially using a GLOB instead of creating the RTD. This is definitely problematic since it doesn't even appear to be linted against, meaning it's very easy to add one in one place and not the other.

SyncIt21 replied:
I could move the static list inside RTD to global_vars and flatten that list here then.  Sure I'll do that

SyncIt21 committed:
Insert New Line

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Merge branch 'master' into RTD

SyncIt21 commented:
base_turf is an list and the depth() function is undefined as of now. 
Might wanna merge ur PR early

Mothblocks commented:
Yeah that's why I said if that PR was merged before yours 😅

SyncIt21 committed:
Merge branch 'master' into RTD

SyncIt21 committed:
Merge branch 'master' into RTD

SyncIt21 committed:
Delete decals & Preserve Overlays

Also Initialize only once

SyncIt21 committed:
Moving Out

1. RTD in its own file
2. tile_rotations in _globvars\lists

SyncIt21 committed:
Semantics

1. Design id infered in client side
2. proper variable names

SyncIt21 committed:
More semantics

1. Tile cost description
2. TIME DEFINES
2. Forgot to change rtd-tgui

SyncIt21 committed:
Merge branch 'master' into RTD

SyncIt21 committed:
Linters

SyncIt21 committed:
Merge branch 'RTD' of https://github.com/SyncIt21/The-TG-Station-Fork into RTD

SyncIt21 committed:
direction & liters

SyncIt21 committed:
Common List

1. Common design list shared by assets & RTD
2. Merged tile directional list with the design list

SyncIt21 committed:
Linters

SyncIt21 committed:
LINTERS!!!

SyncIt21 committed:
Work damn it

SyncIt21 committed:
$&*8__7%

SyncIt21 committed:
Ok now baseturf_at_depth()

SyncIt21 committed:
Don't register duplicate icons

SyncIt21 commented:
And with that icons are the only problem. 
`rotate()` in css works only with angles and not with directions & rotating north_west or in any other combo direction isnt supported either.
For simple directions like north , east etc we can use constant degree values but for directions like north_east or anything else there is no angle to make that possible

SyncIt21 committed:
QDEL i think

SyncIt21 committed:
Proper QDEL() i hope

SyncIt21 committed:
Null check

SyncIt21 committed:
Tile merge

SyncIt21 committed:
End animation

SyncIt21 committed:
Maybe do it manually

SyncIt21 committed:
I'm testing this

SyncIt21 committed:
Testing Testing

SyncIt21 committed:
Ok i jus discovered these macros

Pretty cool

SyncIt21 commented:
For some reason magic carps don't like my RTD. causing my tests to fail
![Screenshot (70)](https://user-images.githubusercontent.com/110812394/209691195-bcd329db-f61f-47b1-9d34-de4b5f873382.png)
It can't be helped

SyncIt21 committed:
Merge branch 'master' into RTD

SyncIt21 committed:
Merge branch 'master' into RTD

Mothblocks commented:
>  rotate() in css works only with angles and not with directions & rotating north_west or in any other combo direction isnt supported either.

i don't understand, is there any reason you can't convert the directions to an angle before passing it to rotate?

Mothblocks commented:
I see no reason to believe any of what you're doing in the BYOND code right now cannot be done in CSS

Mothblocks commented:
I ran some profiling on it and it's not the slowest in the world so I'm fine unblocking later but not until we actually give it a fairer shake than "it takes angles and not directions"

SyncIt21 commented:
> > rotate() in css works only with angles and not with directions & rotating north_west or in any other combo direction isnt supported either.
> 
> i don't understand, is there any reason you can't convert the directions to an angle before passing it to rotate?

Let me explain you what  my css logic to convert direction to rotation is. This is just pseudo code i didnt actually type this but the logic works only to an certain extent as i explain below
 
```
switch(direction)
case NORTH : rotateZ(-180)
break
case SOUTH : <no rotation its the default image>
break
case EAST : rotateZ(90)
break
case WEST : rotateZ(-90)
```

This works for the basic directions perfectly now here are my problems
```
switch(direction)
case NORTH_EAST :
case NORTH_WEST :
case SOUTH_EAST :
case SOUTH_WEST : how do i rotate for these directions?

And that's where my css knowledge ends. I have no idea what to do for these directions
```

SyncIt21 commented:
I fully agree with you to use CSS instead of my wack blending logic but as i said in my previous comment
`How the hell did a tile with only 2 black squares suddenly get 3 black squares by just rotating it`
We need some intresting rotation logic to make it possible

Mothblocks commented:
I didn't even know you could do northeast/southeast/etc tiles. Blegh. Whatever then, undraft if ready. I think the long term solution for this is better lazy asset loading that doesn't need to run on init.

Mothblocks commented:
I didn't even know you could do northeast/southeast/etc tiles. Blegh. Whatever then, undraft if ready. I think the long term solution for this is better lazy asset loading that doesn't need to run on init.

SyncIt21 commented:
> I didn't even know you could do northeast/southeast/etc tiles. Blegh. Whatever then, undraft if ready. I think the long term solution for this is better lazy asset loading that doesn't need to run on init.

I think we can go halfway. Rotate in CSS for the basic directions but for the combo directions use blended icons. That should be possible. Give me an hour

SyncIt21 committed:
Icon optimization

SyncIt21 commented:
Yup this works perfectly. 

Since a lot of tiles rotate in the default NORTH,SOUTH,EAST,WEST directions all these icons are now removed and replaced with simple rotateZ in CSS greatly reducing the final sprite file.

This is very good

SyncIt21 commented:
And now i end every PR with the same comment

SyncIt21 commented:
Done :)

jlsnow301 commented on tgui/packages/tgui/interfaces/RapidTilingDevice.tsx:
```diff
@@ -0,0 +1,181 @@
+const TilePreview = (props, context) => {
+  const { data } = useBackend<Data>(context);
+  const { selected_icon, selected_direction } = data;
+  const rotation = (direction) => {
+    if (direction === 'north') {
+      return 'rotateZ(-180deg)';
+    } else if (direction === 'west') {
+      return 'rotateZ(90deg)';
+    } else if (direction === 'east') {
+      return 'rotateZ(-90deg)';
+    } else {
+      return '';
+    }
+  };
```
Make it a global?
```const ROTATION_MAP = {
    north: 'rotateZ(-180deg)';
    west: 'rotateZ(90deg)';
    east: 'rotateZ(-90deg)';
} as const;
```

then inside TilePreview
`const rotation = ROTATION_MAP[direction] || '';`

SyncIt21 replied:
You sure north doesn't need to be wrapped in single quotes I.e 'north'? direction is of type string

SyncIt21 replied:
No point making it global when it's only used inside this component

SyncIt21 replied:
Getting real late.  I'll do this tom

jlsnow301 replied:
> No point making it global when it's only used inside this component

either way is fine, truly. but this sort of switch/if else chain to make a mapping of strings can be made cleaner

also, not sure why north is highlighted here, it should be ok to use without single quotes.

SyncIt21 replied:
No reason to highlight just wanted to give an example. 
No single quote it is then 
I'll draft this for now so I can properly finish this tom

Mothblocks replied:
it's a bug with using semicolons. github thinks you're doing a new statement

```ts
const ROTATION_MAP = {
    north: 'rotateZ(-180deg)',
    west: 'rotateZ(90deg)',
    east: 'rotateZ(-90deg)',
} as const;
```

jlsnow301 replied:
oh oops! that would do it

SyncIt21 committed:
Merge branch 'master' into RTD

SyncIt21 committed:
Rotation Map

SyncIt21 commented:
Now it's finished

SyncIt21 committed:
Merge branch 'tgstation:master' into RTD

SyncIt21 commented:
Ready for the merge. Anyone intrested?

jlsnow301 commented on tgui/packages/tgui/interfaces/RapidTilingDevice.tsx:
```diff
@@ -0,0 +1,176 @@
+const TilePreview = (props, context) => {
+  const { data } = useBackend<Data>(context);
+  const { selected_icon, selected_direction } = data;
+  const ROTATION_MAP = {
+    north: 'rotateZ(-180deg)',
+    west: 'rotateZ(90deg)',
+    east: 'rotateZ(-90deg)',
+  } as const;
```
Please move this outside of the function

jlsnow301 commented on code/_globalvars/lists/rtd.dm:
```diff
@@ -0,0 +1,116 @@
+			list("name" = "Textured Large", "type" = /obj/item/stack/tile/iron/textured_large, "tile_cost" = 6),
+		),
+		//Looks slightly transparent or faded
+		"Translusent" = list(
```
```suggestion
		"Translucent" = list(
```

SyncIt21 committed:
Review Changes

SyncIt21 committed:
Merge branch 'master' into RTD

SyncIt21 committed:
Reuse RCD icon for worn state cause meh!

SyncIt21 commented:
@jlsnow301 Done 👍

jlsnow301 commented on code/game/objects/items/RTD.dm:
```diff
@@ -0,0 +1,323 @@
+	has_ammobar = TRUE
+	banned_upgrades = RCD_UPGRADE_FRAMES | RCD_UPGRADE_SIMPLE_CIRCUITS | RCD_UPGRADE_FURNISHING
+
+	var/root_category = "Conventional"
+	var/design_category = "Standard"
+	var/datum/tile_info/selected_design
+	var/datum/tile_info/tile_design
+	var/list/design_overlays = list()
+
+/datum/tile_info
+	var/name
+	var/obj/item/stack/tile/tile_type
+	var/icon_state
+	var/cost
+
+	var/ui_directional_data
+	var/tile_directions
+	var/selected_direction
```
These need autodoc comments
Also autodoc comments for procs elsewhere

SyncIt21 committed:
Auto docs

SyncIt21 commented:
@jlsnow301 Done 👍 x 2

jlsnow301 commented on code/game/objects/items/RTD.dm:
```diff
@@ -0,0 +1,336 @@
+#define CONSTRUCTION_TIME 0.4 SECONDS
+#define DECONSTRUCTION_TIME 0.2 SECONDS
+
+/obj/item/construction/rtd
```
Comments here

jlsnow301 commented on code/game/objects/items/RTD.dm:
```diff
@@ -0,0 +1,336 @@
+	has_ammobar = TRUE
+	banned_upgrades = RCD_UPGRADE_FRAMES | RCD_UPGRADE_SIMPLE_CIRCUITS | RCD_UPGRADE_FURNISHING
+
+	var/root_category = "Conventional"
+	var/design_category = "Standard"
+	var/datum/tile_info/selected_design
+	var/datum/tile_info/tile_design
+	var/list/design_overlays = list()
```
Each of these. And so forth

jlsnow301 commented on code/game/objects/items/RTD.dm:
```diff
@@ -0,0 +1,336 @@
+///in short its just an wrapper for mutable appearance where we retrive the nessassary information
+///to recreate an mutable appearance
+/datum/overlay_info
+	var/icon/icon
+	var/icon_state
+	var/direction
+	var/alpha
+	var/color
```
Each

SyncIt21 committed:
Auto Docs Take 2

SyncIt21 commented:
Hope i did this right

jlsnow301 commented on code/game/objects/items/RTD.dm:
```diff
@@ -0,0 +1,352 @@
+
+/// stores the name, type, icon & cost for each tile type
+/datum/tile_info
+	var/name
+	var/obj/item/stack/tile/tile_type
+	var/icon_state
+	var/cost
+
+	var/ui_directional_data
+	var/tile_directions
+	var/selected_direction
```
However redundant, these need autodoc comments

SyncIt21 committed:
Merge branch 'master' into RTD

SyncIt21 committed:
More Auto docs

SyncIt21 commented:
Done

jlsnow301 commented on code/game/objects/items/RTD.dm:
```diff
@@ -0,0 +1,359 @@
+					if(!isnull(floor.managed_overlays))
+						if(islist(floor.managed_overlays))
+							for(var/mutable_appearance/appearance as anything in floor.managed_overlays)
+								design_overlays += new /datum/overlay_info(appearance)
```
A lot of indentation going on here
https://youtu.be/CFRhGnuXG-4

jlsnow301 commented on code/game/objects/items/RTD.dm:
```diff
@@ -0,0 +1,359 @@
+		var/list/target_category =  floor_designs[root_category][sub_category]
+
+		var/list/designs = list() //initialize all designs under this category
+		for(var/i in 1 to target_category.len)
```
Can you use a `for var/category in target_category` loop here?

SyncIt21 replied:
Yup

jlsnow301 commented on code/game/objects/items/RTD.dm:
```diff
@@ -0,0 +1,359 @@
+		qdel(decal)
+
+	//All special effect stuff
+	user.Beam(floor, icon_state="light_beam", time = DECONSTRUCTION_TIME)
```
```suggestion
	user.Beam(floor, icon_state = "light_beam", time = DECONSTRUCTION_TIME)
```

jlsnow301 commented on code/game/objects/items/RTD.dm:
```diff
@@ -0,0 +1,359 @@
+
+	var/floor_designs = GLOB.floor_designs
+	if(!is_valid_plating(floor)) //we infer what floor type it is if its not the usual plating
+		user.Beam(floor, icon_state="light_beam", time = 5)
```
```suggestion
		user.Beam(floor, icon_state = "light_beam", time = 5)
```

jlsnow301 commented on code/game/objects/items/RTD.dm:
```diff
@@ -0,0 +1,359 @@
+		return TRUE
+
+	//All special effect stuff
+	user.Beam(floor, icon_state="light_beam", time = CONSTRUCTION_TIME)
```
```suggestion
	user.Beam(floor, icon_state = "light_beam", time = CONSTRUCTION_TIME)
```

jlsnow301 reviewed (changes requested):
Getting there!

SyncIt21 committed:
null() proc

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

SyncIt21 committed:
Spacing

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

SyncIt21 committed:
More Spacing

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

SyncIt21 committed:
Final Spacing

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

SyncIt21 committed:
More review fixes

SyncIt21 commented:
Need a small help. 

When my player is south facing [i.e., towards the screen] the left & right inhand icons for the RTD is correct.
But when my players is looking left, right, and up the left & right inhand icons uses RCD inhand icon instead of my RTD inhand icon

How do i fix this?

SyncIt21 committed:
Last comment

SyncIt21 committed:
Space

Mothblocks commented:
Looks to me like you didn't change all the directions

![image](https://user-images.githubusercontent.com/35135081/210864501-939f4d34-a971-48c6-9180-4f3808be1528.png)

SyncIt21 commented:
Ahhh crap 

> Looks to me like you didn't change all the directions
> 
> ![image](https://user-images.githubusercontent.com/35135081/210864501-939f4d34-a971-48c6-9180-4f3808be1528.png)

Ahhh crap.  Thank you :)

SyncIt21 committed:
I'm not a very good spriter :(

PAIN!!!

SyncIt21 commented:
Now done
